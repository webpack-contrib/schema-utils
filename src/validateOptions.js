import fs from 'fs';
import path from 'path'; // eslint-disable-line

import Ajv from 'ajv';
import ajvKeywords from 'ajv-keywords';
import ValidationError from './ValidationError';

const ajv = new Ajv({
  useDefaults: true,
  allErrors: true,
  errorDataPath: 'property',
});

ajvKeywords(ajv, ['instanceof', 'typeof']);

const validateOptions = (schema, options, name) => {
  if (typeof schema === 'string') {
    schema = fs.readFileSync(path.resolve(schema), 'utf8'); // eslint-disable-line
    schema = JSON.parse(schema); // eslint-disable-line
  }

  if (!ajv.validate(schema, options)) {
    throw new ValidationError(ajv.errors, name);
  }

  return true;
};

export default validateOptions;
