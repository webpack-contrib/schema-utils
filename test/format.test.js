import validateOptions from '../src';

const FIXTURES = 'test/fixtures/schemas';

describe('Errors', () => {
  const validate = (schema, options) => () =>
    validateOptions(`${FIXTURES}/${schema}.json`, options, '{Name}');

  test('Format - typeof', () => {
    const options = {
      fn: {},
    };

    const err = validate('typeof', options);

    expect(err).toThrowError();
    expect(err).toThrowErrorMatchingSnapshot();
  });

  test('Format - instanceof', () => {
    const options = {
      regexp: {},
    };

    const err = validate('instanceof', options);

    expect(err).toThrowError();
    expect(err).toThrowErrorMatchingSnapshot();
  });
});
