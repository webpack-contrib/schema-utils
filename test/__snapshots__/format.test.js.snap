// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Errors Format - instanceof 1`] = `
"{Name} Invalid Options

options.regexp should be {RegExp}
"
`;

exports[`Errors Format - typeof 1`] = `
"{Name} Invalid Options

options.fn should be a {Function}
"
`;
